docs/gguf.md

```diff
@@ -0,0 +1,382 @@
+
+### General
+
+- **`general.architecture: string`**: describes what architecture this model implements. All lowercase ASCII, with only `[a-z0-9]+` characters allowed. Known values include:
+  - `llama`
+  - `mpt`
+  - `gptneox`
+  - `gptj`
+  - `gpt2`
+  - `bloom`
+  - `falcon`
+  - `rwkv`
+- **`general.quantization_version: u32`**: version of quantization scheme
+- `general.file_type: string`: type of the majority of the tensors in the file. This shouldn't have any semantic meaning and should be purely informational, hence the use of `string`.
+- `general.license: string`: SPDX license of the model
+- `general.description: string`: free-form description of the model including anything that isn't covered by the other fields
+- `general.source.url: string`: URL to the source of the model. Can be a GitHub repo, a paper, etc.
+- `general.source.huggingface.repository: string`: Hugging Face model repository that this model is either hosted on or based on
+
+### LLM
+
+In the following, `[llm]` is used to fill in for the name of a specific LLM architecture. They will be used in each architecture's section.
+
+- `[llm].context_length: u32`: size of the maximum supported context
+- `[llm].hidden_size: u32`: embedding layer size
+- `[llm].num_layers: u32`: number of layers
+- `[llm].num_ff: u32`: The length of the feedforward layer.
+- `[llm].use_parallel_residual: bool`: whether or not the parallel residual logic should be used
+- `[llm].max_seq_len: u32`: Maximum sequence length
+- `[llm].attention.num_heads: u32`: number of attention heads
+- `[llm].attention.alibi_bias_max: f32`: The maximum bias to use for ALiBI
+- `[llm].attention.clip_kqv: f32`: Value (`C`) to clamp the values of the `Q`, `K`, and `V` tensors between (`[-C, C]`).
+- `[llm].rope.num_dims: u32`: The number of rotary dimensions for RoPE.
+
+#### Models
+
+The following sections describe the metadata for each model architecture. Each key specified _must_ be present.
+
+##### LLaMA
+
+- `llama.context_length`
+- `llama.hidden_size`
+- `llama.num_layers`
+- `llama.num_ff`
+- `llama.rope.num_dims`
+- `llama.attention.num_heads`
+
+##### MPT
+
+- `mpt.max_seq_len`
+- `mpt.hidden_size`
+- `mpt.num_layers`
+- `mpt.attention.num_heads`
+- `mpt.attention.alibi_bias_max`
+- `mpt.attention.clip_kqv`
+
+##### GPT-NeoX
+
+- `gptneox.context_length`
+- `gptneox.hidden_size`
+- `gptneox.num_layers`
+- `gptneox.use_parallel_residual`
+- `gptneox.rope.num_dims`
+- `gptneox.attention.num_heads`
+
+##### GPT-J
+
+- `gptj.context_length`
+- `gptj.hidden_size`
+- `gptj.num_layers`
+- `gptj.rope.num_dims`
+- `gptj.attention.num_heads`
+
+##### GPT-2
+
+- `gpt2.context_length`
+- `gpt2.hidden_size`
+- `gpt2.num_layers`
+- `gpt2.attention.num_heads`
+
+##### BLOOM
+
+- `bloom.context_length`
+- `bloom.hidden_size`
+- `bloom.num_layers`
+- `bloom.num_ff`
+- `bloom.attention.num_heads`
+
+##### Falcon
+
+**TODO**.
+
+##### RWKV
+
+**TODO**.
```

Comment:
I propose the following:

- `rwkv.rwkv_architecture_version`, the only allowed value currently is `4`, will be `5` in the future
- `rwkv.context_length`, recommended maximum context length
- `rwkv.num_layers`, counting only RWKV+FFN layers, excluding `embedding` and `head` -- same as `n_layer` in [rwkv.cpp file format](https://github.com/saharNooby/rwkv.cpp/blob/master/docs/FILE_FORMAT.md)
- `rwkv.hidden_size` -- same as `n_embed` in rwkv.cpp file format
- `rwkv.vocab_size` -- same as `n_vocab` in rwkv.cpp file format
- `rwkv.ffn_dimension` -- dimension of FFN sub-layer, defined as the height of `blocks.0.ffn.key.weight`

Reply:
I think @BlinkDL and @LoganDark may also want to have a say in this :)

Reply:
> `rwkv.rwkv_architecture_version`

why not `rwkv.version`?

> `rwkv.hidden_size`

this is misleading, I would just do `rwkv.embed_side` because this is the width of the `emb`ed tensor

> `rwkv.ffn_dimension`

I don't think `dimension` would be a good word to use here but not sure what to use instead

Reply:
Agree with Logan based on what I know of RWKV, but will let y'all hash it out first

Reply:
@LoganDark 

>why not rwkv.version?

To avoid confusion with other versions, like file format version or quantization format version. It just specifies clearly what exactly is versioned here. I think it can be simplified to `rwkv.architecture_version` tho.

>this is misleading, I would just do rwkv.embed_size

I've chosen it because above in "LLM" section following key is recommended: `[llm].hidden_size: u32: embedding layer size`. Having "embed_size" for RWKV only would be pretty inconsistent, while all other models use "hidden_size" for the same thing.

Reply:
Hm, there's a generic proposed parameter `[llm].n_ff: u32: the length of the feedforward layer` for FFN layer size. WDYT on using it as `rwkv.n_ff`?

(For me personally, looks too short and cryptic)